src/inmanta/server/services/compilerservice.py

```diff
@@ -213,37 +218,61 @@ async def run(self, force_update: Optional[bool] = False) -> Tuple[bool, Optiona
+                await self._run_compile_stage(
+                    "Installing modules", [*inmanta_path, "-vvv", "-X", "project", "install"], project_dir
+                )
+
             repo_url: str = env.repo_url
             repo_branch: str = env.repo_branch
-            if not repo_url:
-                if not os.path.exists(os.path.join(project_dir, "project.yml")):
-                    await self._warning(f"Failed to compile: no project found in {project_dir} and no repository set")
-                await self._end_stage(0)
-            else:
+            if os.path.exists(os.path.join(project_dir, "project.yml")):
                 await self._end_stage(0)
-                # checkout repo
-
-                if not os.path.exists(os.path.join(project_dir, ".git")):
-                    cmd = ["git", "clone", repo_url, "."]
-                    if repo_branch:
-                        cmd.extend(["-b", repo_branch])
-                    result = await self._run_compile_stage("Cloning repository", cmd, project_dir)
-                    if result.returncode is None or result.returncode > 0:
-                        return False, None
-
-                elif force_update or self.request.force_update:
-                    result = await self._run_compile_stage("Fetching changes", ["git", "fetch", repo_url], project_dir)
+
+                should_update: bool = force_update or self.request.force_update
+
+                # switch branches
                 if repo_branch:
                     branch = await self.get_branch()
                     if branch is not None and repo_branch != branch:
+                        if should_update:
+                            await self._run_compile_stage("Fetching new branch heads", ["git", "fetch"], project_dir)
                         result = await self._run_compile_stage(
-                            f"switching branch from {branch} to {repo_branch}", ["git", "checkout", repo_branch], project_dir
+                            f"Switching branch from {branch} to {repo_branch}", ["git", "checkout", repo_branch], project_dir
                         )
+                        if not should_update:
+                            # if we update, update procedure will install modules
+                            await install_modules()
 
-                if force_update or self.request.force_update:
+                # update project
+                if should_update:
                     await self._run_compile_stage("Pulling updates", ["git", "pull"], project_dir)
-                    LOGGER.info("Installing and updating modules")
-                    await self._run_compile_stage("Updating modules", inmanta_path + ["modules", "update"], project_dir)
+                    await update_modules()
+            else:
+                if not repo_url:
+                    await self._warning(f"Failed to compile: no project found in {project_dir} and no repository set.")
```

Comment:
A remark that is not linked to this pull request: I think that the logic in the `finally` block at the end of this method, should log to `self._warning()` instead of using the standard `LOGGER`. Do you agree?

Reply:
I agree, mostly. I've updated this to use `self._warning` if a stage is active.